Fix bugs in in_memory batches processing

MelLain opened the pull request.

MelLain commented:
Fix releated to issue-804

JeanPaulShapo commented:
@MelLain Just use sharp sign (#) before issue number to link it :)

JeanPaulShapo commented on python/artm/wrapper/spec.py:
```diff
@@ -94,7 +95,7 @@ def __init__(self, name, arguments, result=None, request=None):
         'ArtmDisposeBatch',
-        [('master_id', int), ('name', str)],
+        [('master_id', int), ('name', unicode if sys.version_info.major == 2 else str)],
     ),
```
I think it's better to use `six` library here (i.e [text_type](https://pythonhosted.org/six/index.html#six.text_type)).

MelLain replied:
ok, I'll try it

JeanPaulShapo commented on python/tests/artm/test_batches_utils.py:
```diff
@@ -55,6 +55,7 @@ def test_func():
         model.fit_online(update_every=1, batch_vectorizer=in_memory_batch_vectorizer)
         assert len(model.score_tracker['perplexity'].value) == 14
 
+        del in_memory_batch_vectorizer
```
Do you think it's better to comment what functionality is supposed to test in this line?

MelLain replied:
No, bv evidently should be deletable

JeanPaulShapo replied:
Sorry, it was ambiguous question :)
I'd like to say 'Is it better to add comment before this line explaining what you want to test?' :)

MelLain replied:
No, I don't think that it's necessary)

JeanPaulShapo replied:
ok

JeanPaulShapo commented on python/artm/artm_model.py:
```diff
@@ -929,7 +929,9 @@ def transform(self, batch_vectorizer=None, theta_matrix_type='dense_theta',
+        batches_list = batch_vectorizer.batches_list
+        if not batch_vectorizer.process_in_memory:
+            batches_list = [batch.filename for batch in batch_vectorizer.batches_list]
 
```
Looks good! 👍
Anyway, I have an idea about adding property in `BatchVectorizer` class which returns filenames or ids according to type of processing. Can it be better to hide `process_in_memory` state from external environment?

MelLain replied:
Maybe, I'll think about it

JeanPaulShapo commented on python/artm/batches_utils.py:
```diff
@@ -120,15 +120,15 @@ def __init__(self, batches=None, collection_name=None, data_path='', data_format
+        if self._remove_batches:
+            shutil.rmtree(self._target_folder)
+        self._remove_batches = False
+
```
Could you give me some hints why order matters?

MelLain replied:
It more convenient for me, thats the only reason)

JeanPaulShapo replied:
Ok, no problem :)

JeanPaulShapo commented on python/artm/master_component.py:
```diff
@@ -888,7 +888,8 @@ def import_batches(self, batches=None):
         """
         if batch_id is not None:
-            self._lib.ArtmDisposeBatch(self.master_id, batch_id)
+            self._lib.ArtmDisposeBatch(self.master_id, u'{}'.format(batch_id))
```
There are some problems with converting unusual str-like objects to unicode in Python.
E.g. try to execute the following line:
`
unicode_sample = u'{}'.format('пример')
`

JeanPaulShapo replied:
It may be better to accept only unicode-like objects. Or not (I don't know)?

MelLain replied:
You're right, it would be better to use only unicode

MelLain commented:
thanks, I didn't know, how to link to issue)

#806 
#804 

no, it is a full fix that can be merged with closing issue.

JeanPaulShapo commented:
Ok, I see.

JeanPaulShapo commented:
Feel free to merge after passing CI tests :)

JeanPaulShapo commented:
@MelLain There are some problems with PEP8 in Travis CI:
`python/artm/artm_model.py:926:121: E501 line too long (121 > 120 characters)`

JeanPaulShapo commented on python/artm/batches_utils.py:
```diff
@@ -276,9 +276,13 @@ def __reset_batch():
     @property
     def batches_list(self):
         """
-        :return: list of batches names
```
@MelLain Unfortunately, right now changing logic of `batches_list` property breaks code which uses it rather weird way, e.g. `hARTM` :( In my opinion, right now it's ok to use simple hack with `process_in_memory` until we understand problem with `hARTM` class

AlexeySanko commented:
Hi.
I understand that I'm not in time, but while changes into memory processing are in progress:
Why did you use additionally `process_in_memory` argument for BatchVectorizer.__init__?
Common way for this function as I can see is data_format usage and we can just add new one - "memory_batch".
In this case we will avoid different type comparisons into conditions like:
[https://github.com/bigartm/bigartm/pull/806/files#diff-8ac32600e1bf043123411a7dc9657e48R89](elif data_format == 'batches' or self._process_in_memory:)
Thanks.

MelLain commented:
The problem with hartm is that we need to deal with batches as objects in it, so I decided to separate batches objects and their identifiers.

MelLain commented:
@AlexeySanko 

`or self._process_in_memory:` - this check is useless, I removed it.

Also I removed `process_in_memory` flag, because it's enough to set the `model` parameter of BV to use it in "in memory" mode.

The reason I separated data_format and proc in mem flag is that proc in memory not a format, it's only the way to deal with given batches (currently from disk).

@JeanPaulShapo please, make a full review of all changes again.

JeanPaulShapo commented:
Hi, @AlexeySanko.
Your suggestion looks good.
However, proposed change is likely to be backward-incompatible, because you want to remove some existing functionality from API.
I think it'll be better if you open issue about this change, and we consider it when making new major version.

MelLain commented:
It is not a large api breaking change, so it can be safely pushed in new release.

JeanPaulShapo commented:
@MelLain
OK, I agree.
Is it better to add some description about it in Release Notes (in this PR, apparently)?

MelLain commented:
Yes, we only need to say, that in memory flag is deprecated and should be simply deleted in all code cases.

JeanPaulShapo commented:
@MelLain Great!

AlexeySanko commented:
Hi, @MelLain.
My outside stupid view says that "batch from disk" and "batch from memory" are different formats. :) Because first are just files and second - Python list (I can create it manually without any file usage). I understand that we get the same batch list finally, but now we're talking about third-party user who doesn't know.

Now there is hidden logic: `self._process_in_memory = data_format == 'batches' and model is not None`
And arguments list do not provide clear understanding how to force in-memory processing.
Rename `model` argument to `process_in_memory_model`?
Add "in_memory_batch" data_format and make first row data_format changing for keeping similar processing for in-memory and not?
`if data_format == "in_memory_batch":
   self._process_in_memory = True
   data_format = "batch"
`

Doesn't look like very beautiful, but will make user's life easier.

Thanks.

MelLain commented:
@AlexeySanko I think, that `process_in_memory_model` renaming is the best solution)

AlexeySanko commented:
@MelLain, ok. Thanks!

AlexeySanko commented on python/artm/batches_utils.py:
```diff
@@ -73,17 +72,18 @@ def __init__(self, batches=None, collection_name=None, data_path='', data_format
-            raise IOError("Correct configuration: process_memory == True + data_format == 'batches' + model != None")
+        self._process_in_memory = data_format == 'batches' and process_in_memory_model is not None
+        if not self._process_in_memory and process_in_memory_model is not None:
+            raise IOError("Correct configuration for in memory processing: data_format ==" +\
```
Into brackets Python provides string concatenation without + and \
```
            raise IOError("Correct configuration for in memory processing: data_format =="
                          "'batches' + process_in_memory_model != None")
```

JeanPaulShapo commented on docs/release_notes/python.txt:
```diff
@@ -8,6 +8,10 @@ For this reason we recommend to review :doc:`messages`.
 
+v0.8.4
+------
+* Flag ``process_in_memory`` was removed from ``artm.BatchVectorizer``, and ``model`` parameter ws renames into ``process_in_memory_model``. Now specification of this parameter with ``data_format='batches'`` will mean in memory processing of given in ``batches`` parameter batches.
```
Try to write in *imperative mood* like `Remove flag ... from ..., and rename ... to ...`

JeanPaulShapo replied:
*in memory* -> *in-memory* (example *in-memory database*)

JeanPaulShapo replied:
I think, it'll be better to change *processing of given in ``batches`` parameter batches* to *processing of batches, specified in ``batches`` parameter*

JeanPaulShapo replied:
*Now specification of this parameter with ... will mean ...* -> *Passing this argument with ... triggers ...*

JeanPaulShapo commented on python/artm/artm_model.py:
```diff
@@ -582,10 +574,10 @@ def _func(batch):
                                          update_after_final, apply_weight_final,
                                          decay_weight_final, async)),
-            len(batches_list))
+            len(batch_vectorizer.batches_ids))
```
Is it OK to write `__len__` function to count number of entities in `batch_vectorizer` instead `num_batches` property and even ugly `len(batch_vectorizer.batches_ids)`, like in file `Scores` class from `scores.py`?

MelLain replied:
I agree, that len(...) should be replaces with method call, but I don't think, that `__len__` is a good idea, because `num_batches` is more convenient.

JeanPaulShapo commented on python/artm/batches_utils.py:
```diff
@@ -73,17 +72,18 @@ def __init__(self, batches=None, collection_name=None, data_path='', data_format
-        :param artm.ARTM model: ARTM instance that will use this vectorizer, is required when\
-                                process_in_memory == True. Will be ignored in other cases
+        :param artm.ARTM process_in_memory_model: ARTM instance that will use this vectorizer, is\
+                                                  required when one need processing of batches from\
```
*one need**s** *?

JeanPaulShapo commented on python/artm/batches_utils.py:
```diff
@@ -73,17 +72,18 @@ def __init__(self, batches=None, collection_name=None, data_path='', data_format
+        :param artm.ARTM process_in_memory_model: ARTM instance that will use this vectorizer, is\
+                                                  required when one need processing of batches from\
+                                                  disk in RAM (only if data_format == Batches).\
+                                                  NOTE: makes vectorizer model specific.
```
lower `Batches` and enclose it in quotes

JeanPaulShapo commented:
Fixes #804
